docs: Add plotly.js v4 migration guide (#455)
plotly/graphing-library-docs: _posts/plotly_js/guides/migrating-to-v4/2026-07-17-migrating-to-v4.md

Most figures will render without changes. The items below call out where
existing code or input needs to be updated.

## Contents

- [Minimum Node.js version](#minimum-nodejs-version)
- [Removed color inputs](#removed-color-inputs)
- [Removed Chart Studio APIs](#removed-chart-studio-apis)
- [Removed mapbox traces and subplots](#removed-mapbox-traces-and-subplots)
- [Country name lookup](#country-name-lookup)
- [Auto-fitting on `map` and `geo` subplots](#auto-fitting-on-map-and-geo-subplots)
- [Scattermap rendering changes](#scattermap-rendering-changes)
- [Shape legend `line.dash`](#shape-legend-line-dash)
- [Overlaying axis `tickmode`](#overlaying-axis-tickmode)
- [`splom.axis.matches` default](#splomaxismatches-default)
- [Sankey layout](#sankey-layout)
- [MathJax v2 dropped](#mathjax-v2-dropped)

---

## Minimum Node.js version

Plotly.js now requires Node.js v22 or newer. Upgrade if you build from source,
run Plotly in a Node runtime (SSR via jsdom, image export pipelines,
validation scripts), or use `yarn` / `pnpm` (which enforce `engines`
strictly).

Consumers of the pre-built browser bundle are unaffected.

---

## Removed color inputs

> Three color-input formats no longer work — invalid inputs fall back to the

Reviewer: Plotly.js colors now conform thoroughly to the CSS standard {link}. As a result...

Author: Updated.

attribute's default:

| Input | What to do |
|---|---|
| `hsv()` strings | Convert to `hsl()`, `hwb()`, hex, or `rgb()` |
| `rgb()` / `rgba()` strings with 0–1 decimal fractions | Convert `'rgb(0.5, 0.5, 0.5)'` to `'rgb(128, 128, 128)'` (or any supported string form) |

> | Numeric color values (e.g. `color: 42`) | Convert to a color string |

Reviewer: What did `color: 42` do before? Should make it clear that this does NOT apply to situations using a color map. `color: [1,2,3,4,5]` still works as it always did.

Author: That color was actually rejected before as well, so it fell back to the default plotly colors. `tinycolor2` will return black with the value, but `color` will throw. I got the bare library behavior mixed up with how plotly handles things. I'll remove it.

| `scrollZoom: 'mapbox'` (e.g. `'mapbox+cartesian'`) | `'map'` |

The attribute shape is unchanged — usually a straight rename works:

```js
// Before
{
    data: [{ type: 'scattermapbox', lon: [...], lat: [...] }],
    layout: { mapbox: { style: 'open-street-map', center: {...}, zoom: 2 } },
    config: { mapboxAccessToken: 'pk.eyJ1...' },
}

// After
{
    data: [{ type: 'scattermap', lon: [...], lat: [...] }],
    layout: { map: { style: 'open-street-map', center: {...}, zoom: 2 } },
    // mapboxAccessToken removed
}
```

Built-in style names still work but use different tile providers (Carto,
ArcGIS, OpenStreetMap) — visual output will shift. Stamen styles
(`stamen-terrain`, `-toner`, `-watercolor`) are no longer built in — they now
require a Stadia API key. The `mapbox://styles/mapbox/...` short-form URL is
no longer resolved.

---

## Country name lookup

`locationmode: 'country names'` on `choropleth` and `scattergeo` traces now
uses a stricter matcher. Strings that only matched previously via broad regex
fragments (e.g. `'Republic of'`, `'Not Iran'`) no longer resolve — locations
are skipped. Switch to a canonical country name, an alias, or an ISO-3 / ISO-2

> code.

Reviewer: Link out to each of these sets - canonical names, full list of aliases, and standard reference lists for ISO-3 and ISO-2

Author: I added a link to the source for the library.

maintained and supports modern CSS Color 4 syntax, but a handful of
behaviors changed in the process.

### Color string inputs that no longer work

**`hsv()` color strings are no longer accepted.**

```js
// Before
{ marker: { color: 'hsv(200, 80%, 80%)' } }

// After — convert to hsl, hwb, hex, or rgb
{ marker: { color: 'hsl(200, 67%, 47%)' } }
```

A color in the unsupported format will fall back to the attribute's default.

**`rgb()` / `rgba()` strings with 0–1 decimal fractions are no longer
accepted.**

```js
// Before — tinycolor parsed this as 0–1 fractions
'rgb(0.5, 0.5, 0.5)'  // → rgb(128, 128, 128)

// After — the underlying library reads them as integer 0–255 (truncated)
'rgb(0.5, 0.5, 0.5)'  // → rgb(1, 1, 1)
```

Convert any inputs of the form `rgb(0.5, 0.5, 0.5)` to `rgb(128, 128, 128)`
(or any other supported color string) before passing them to Plotly.

Mixed-format strings where at least one component is `>= 1` (e.g.
`rgb(255, 0.0, 0.0)`) continue to render correctly: the new parser treats
the `0.0` components as integer `0`.

**Numeric color inputs are now rejected.**

```js
// Before — silently accepted, rendered as near-black
{ marker: { color: 42 } }

> // After — falls back to the attribute default

Reviewer 2: ```suggestion
```

'hsla(0, 100%, 50%, 0.5)' // hsla
'hwb(0, 0%, 0%)'     // hwb (hue-whiteness-blackness)
```

See [test/image/mocks/color_syntax_formats.json](https://github.com/plotly/plotly.js/blob/master/test/image/mocks/color_syntax_formats.json)
for a working example using each format.

#### Browser compatibility

All of these are parsed by Plotly in JavaScript and normalized to legacy
`rgb(...)` / `rgba(...)` before being written to the DOM, so the browser
never sees the modern syntax — your figures render the same in any
browser that ran v3. The note below only matters if you also mirror these
strings into your own stylesheets or template literals.

Native CSS support:

| Format | First supported |
|---|---|
| `#rrggbbaa`, `#rgba` | Chrome 62 (2017), Firefox 49 (2016), Safari 9.1 (2016) |
| `rgb(r g b)`, `rgb(r g b / a)`, `hsl(h s l / a)` (space + slash-alpha) | Chrome 65 (2018), Firefox 52 (2017), Safari 12.1 (Mar 2019) |
| `hsla()` | Universal since IE9 |
| `hwb()` | Chrome 101 (Apr 2022), Firefox 96 (Jan 2022), Safari 15 (Sep 2021) |

`hwb()` is the only format here that won't parse in browsers from ~2020
or earlier. Everything else is safe in any browser shipped in the last

> seven years.

Reviewer 2: I'm not sure this level of detail is needed in the migration guide, could it be moved to a docs page on color strings instead?
```suggestion
```

| `layout.mapbox.accesstoken` | None — MapLibre uses open tile providers |
| `mapboxAccessToken` config option | None — see above |
| `plotly-mapbox.min.js` partial bundle | `plotly-map.min.js` |
| Modebar buttons `zoomInMapbox`, `zoomOutMapbox`, `resetViewMapbox` | `zoomInMap`, `zoomOutMap`, `resetViewMap` |
| `scrollZoom` flag value `'mapbox'` (e.g. `scrollZoom: 'mapbox+cartesian'`) | `'map'` |

The schema for `map` mirrors `mapbox` attribute-for-attribute (`domain`,
`style`, `center`, `zoom`, `bearing`, `pitch`, `bounds`, `layers[...]`),
so the migration is usually a rename.

```js
// Before
{
    data: [{ type: 'scattermapbox', lon: [...], lat: [...] }],
    layout: {
        mapbox: {
            style: 'open-street-map',
            center: { lon: 0, lat: 0 },
            zoom: 2,
        },
    },
    config: { mapboxAccessToken: 'pk.eyJ1...' },
}

// After
{
    data: [{ type: 'scattermap', lon: [...], lat: [...] }],
    layout: {
        map: {
            style: 'open-street-map',
            center: { lon: 0, lat: 0 },
            zoom: 2,
        },
    },
    // mapboxAccessToken removed — MapLibre needs no token for the
    // built-in styles
}
```

> ### Built-in style values

Reviewer 3: Isn't this previously covered when we introduced the maplibre traces? Could we reference an existing doc in the "dropping map box" section?

Author: I'll add a [link](https://plotly.com/python/mapbox-to-maplibre/).

when no other figure changes:

| Style name | Old (`mapbox`) | New (`map`) |
|---|---|---|
| `basic`, `streets`, `outdoors` | mapbox.com vector tiles (required token) | Carto Voyager |
| `light` | mapbox.com light | Carto Positron |
| `dark` | mapbox.com dark | Carto Dark Matter |
| `satellite`, `satellite-streets` | mapbox.com satellite | ArcGIS World Imagery |
| `white-bg`, `open-street-map` | unchanged | unchanged |
| `carto-*`, `carto-*-nolabels` | n/a (already open) | unchanged |

The three Stamen styles (`stamen-terrain`, `stamen-toner`,
`stamen-watercolor`) are gone — Stamen tiles now require a Stadia API
key. Users who need them can supply a custom `style` URL.

### Custom style URLs

Custom MapLibre / Mapbox-style JSON URLs continue to work. The
`mapbox://styles/mapbox/...` short form is no longer resolved, since it
requires authenticating to mapbox.com. Substitute the full URL of an
equivalent open style, or self-host the style JSON.

### Static export

The Mapbox logo previously injected into static SVG / PNG exports of
mapbox subplots is no longer drawn — MapLibre has no equivalent
trademark requirement. Attribution provided by tile sources (Carto,

> OpenStreetMap, ArcGIS) is still rendered inside the map canvas.

Reviewer 2: ```suggestion
```

---

## Country name lookup

*Implemented in [#7856](https://github.com/plotly/plotly.js/pull/7856).*

`locationmode: 'country names'` on `choropleth` and `scattergeo` traces is
now resolved by [country-iso-search](https://github.com/plotly/country-iso-search)
instead of [country-regex](https://github.com/etpinard/country-regex). The
attribute name, accepted values list, and ISO-3 output are unchanged —
the difference is in *how* a name string is matched to an alpha-3 code.

The old library walked a table of country-specific regular expressions
and returned the first match. That made it lenient (substring matches
worked) but ambiguity-prone — strings like `'Republic of'` could
resolve to whichever entry's regex hit first. The new library does a
sanitized exact lookup against a curated alias table, which is stricter
about ambiguous input but accepts many more legitimate forms.

> ### Names that resolved before but no longer do

Reviewer 3: This title isn't specific enough, it sounds like a Kafka sentence 🙃

Author: _The real name is friendship._

I'll update it.

("FullData") is internal.

**`PlotType` is an alias for `TraceType`.** Both name the same
string-literal union of trace type names; existing code that imports
`PlotType` keeps compiling. New code should prefer `TraceType`.

---

## Scattermap icon defaults

*Implemented in [#7825](https://github.com/plotly/plotly.js/pull/7825).*

If you used `scattermap` or `scattermapbox` in v3, three rendering details
have changed by the time you land on `scattermap` in v4:

### Marker icons now respect `marker.color`

`marker.color` on a scattermap trace is now applied to the icon by
setting `icon-color` (plus a small `icon-halo` so the colored edges
read cleanly). In v3 the property was set on the layer but the
underlying icons weren't SDF sprites, so the color had no visible
effect — coloring was effectively a no-op.

In v4 the Maki icons are loaded with `{sdf: true}` and tinted with
`marker.color`. Figures relying on the icon retaining its original
artwork color will now render in the trace color. If you need the
original tint, pick a Maki icon already in that color and set
`marker.color: 'white'` (which becomes the identity tint for SDF

> sprites).

Reviewer 2: Too much detail

```suggestion
`marker.color` on a scattermap trace is now applied to the icon by
setting `icon-color`.
```

*Implemented in [#7884](https://github.com/plotly/plotly.js/pull/7884).*

`map` subplots now auto-frame the view to fit the trace data. In v3, the
same layout produced a world view centered at `(0, 0)` with `zoom: 1`,
requiring the user to supply `center` and `zoom` (or drag/scroll) to see
their data.

```js
Plotly.newPlot(gd, [{
    type: 'scattermap',
    lon: [-122.4, -73.9, -87.6],
    lat: [37.8, 40.7, 41.9]
}], {
    map: { style: 'basic' }
});
// v3: map opens centered at (0, 0) at zoom 1
// v4: map opens framed on the three cities
```

| Before (v3) | After (v4) |
|---|---|
| ![v3 map with world view](/all_static/images/migrating-to-v4/migration_map_autofit_v3.png) | ![v4 map framed on cities](/all_static/images/migrating-to-v4/migration_map_autofit_v4.png) |

Auto-framing runs on the initial render and again on data-only updates
(`Plotly.restyle` of `lon`/`lat`, `Plotly.addTraces`, `Plotly.deleteTraces`)
whenever the user hasn't overridden the view. Once the user pans, zooms,
rotates, or tilts the map — or explicitly sets `center` / `zoom` / `bearing`
/ `pitch` in the layout — auto-framing steps aside and preserves the
chosen view across further data changes.

> ### Opting out

Reviewer 3: Should be one level deeper

> ### Opting out

Reviewer 2: ```suggestion
### Disabling auto-framing
```

> Two ways to disable auto-framing.

Reviewer 2: ```suggestion
There are two ways to disable auto-framing.
```

**Set `map.fitbounds: false`** — recommended for intentional opt-out;

> explicit and independent of the view values.

Reviewer 2: ```suggestion
**Set `map.fitbounds: false`** — disables automatic auto-framing and instead defaults to showing a map with center `(0, 0)` and zoom `1`, as in v3.
```

```js
{ map: { fitbounds: false } }
```

**Set `map.center` or `map.zoom` explicitly** — any value counts as

> user intent, including `zoom: 0` and `center: { lon: 0, lat: 0 }`.

Reviewer 2: ```suggestion
**Set `map.center` or `map.zoom` explicitly** — passing any value for `map.zoom` or `map.center`, (including 0), uses the provided values rather than choosing zoom and center via auto-framing.
```

```js
// Keep the v3 world view:
{ map: { center: { lon: 0, lat: 0 }, zoom: 1 } }
```

### Which traces contribute

Each `map` subplot auto-frames independently, using only the
`scattermap` and `densitymap` traces assigned to it (via the trace's
`subplot` attribute). Each contributing trace's `lon` / `lat` arrays
feed the frame, with non-finite entries skipped. Traces belonging to
a different subplot (`map2`, `polar`, etc.) are framed by their own
subplot, not this one; traces with `visible !== true` are ignored
entirely.

**`choroplethmap` traces skip auto-framing** because they carry
`locations` / `geojson` rather than point lon/lat, so their extent
can't be included without geojson bbox handling. When a visible
`choroplethmap` sits on the subplot, the view falls back to the v3

> default. Set `map.center` and `map.zoom` explicitly to frame the choropleth.

Reviewer 2: ```suggestion
### Which traces are affected

`scattermap` and `densitymap` traces are affected by this change. **`choroplethmap` traces skip auto-framing** because their data contains geographic regions, whose bounding box must be computed differently. When a subplot contains a `choroplethmap` trace, auto-framing is disabled and the v3 default of center `(0, 0` and zoom `1` is used.
```

### Antimeridian data

For data that straddles the antimeridian (points on both sides of ±180°),
the auto-frame picks the compact crossing range rather than the
long-way-around view. For example, `lon: [131.8855, -179]` frames as
`[131.8855, 181]` (a ~49° span across the antimeridian) instead of the

> naive `[-179, 131.8855]` (~311° the wrong way).

Reviewer 2: ```suggestion
### Data that crosses the antimeridian

For data that straddles the antimeridian (points on both sides of ±180°, in the middle of the Pacific Ocean),
the auto-frame picks the compact crossing range rather than the
long-way-around view. For example, `lon: [131.8855, -179]` frames as
`[131.8855, 181]` (a ~49° span across the antimeridian) instead of the
naive `[-179, 131.8855]` (~311° the wrong way).
```
Would be nice to have a visual example here

### `map.bounds` is unaffected

The `map.bounds` layout attribute (which restricts panning) is a
separate concern and continues to work as in v3. Auto-framing computes
the *initial* view; `bounds` constrains *interactive* view movement.

> Both can coexist.

Reviewer 2: ```suggestion
```

a straight on/off toggle, while `geo.fitbounds` also selects *what*
contributes to the frame (visible `locations` only, vs. the entire trace
`geojson`).

---

## Geo subplots auto-frame by default

*Implemented in [#7895](https://github.com/plotly/plotly.js/pull/7895).*

`layout.geo.fitbounds` now defaults to `'locations'` (v3 default: `false`).
`geo` subplots — used by `scattergeo`, `choropleth`, and any other
location-based geo trace — now auto-frame their initial view to fit the
trace data, matching the behavior of `map` subplots. Figures that previously
rendered the full world map (or the scope's default extent) now zoom to just
the locations they plot.

```js
Plotly.newPlot(gd, [{
    type: 'choropleth',
    locations: ['CAN', 'MEX'],
    z: [1, 2]
}], {
    // v3: full world map (fitbounds defaulted to false)
    // v4: tight fit around Canada and Mexico (fitbounds now defaults to 'locations')
});
```

| Before (v3) | After (v4) |
|---|---|
| ![v3 geo world view](/all_static/images/migrating-to-v4/migration_geo_autofit_v3.png) | ![v4 geo framed on Canada+Mexico](/all_static/images/migrating-to-v4/migration_geo_autofit_v4.png) |

> ### How to opt out

Reviewer 2: ```suggestion
### Disabling auto-framing
```

> Two paths:

Reviewer 2: ```suggestion
There are two ways to disable auto-framing.
```

> **Explicit:** set `layout.geo.fitbounds: false` to restore the v3 default view.

Reviewer 2: ```suggestion
**Set `layout.geo.fitbounds: false`** — disables automatic auto-framing and instead defaults to showing a map with center `(0, 0)` and zoom `1`, as in v3.
```

```js
Plotly.newPlot(gd, [/* ... */], {
    geo: { fitbounds: false }  // world view / scope-default view, as in v3
});
```

**Implicit:** setting any of the view attributes that auto-fit would
otherwise compute automatically disables auto-fit for that figure. Which
attributes count depends on the projection:

| Projection kind | Projection-specific opt-out attributes |
|---|---|
| Any non-`world` scope (checked first, even if the projection is also clipped) | `center.lon`, `center.lat`, `projection.scale` |
| World scope + clipped projection (orthographic, azimuthal equal area, etc.) | `center.lon`, `center.lat`, `projection.scale`, `projection.rotation.lon`, `projection.rotation.lat` |

> | World scope, non-clipped projection (miller, robinson, mercator, etc.) | `center.lon`, `center.lat`, `projection.scale`, `projection.rotation.lon` |

Reviewer 2: The only reason this differs across projections is because different projections support different view attributes, right?

Could we say something more general like "setting `layout.center` or any value under `layout.projection` disables auto-fitting. See <docs page> for a summary of which projection attributes are supported by which geo projections.

Author: That's correct. I'll update it.

### Projections that don't participate in auto-fit

Three projection types skip auto-fit and render with the schema default
view even when `fitbounds` is unset (i.e., defaulted to `'locations'`):

- `'albers usa'` — a fixed composite projection with predetermined insets
  for Alaska and Hawaii; it has no `center`/`rotation` concept for fitbounds
  to target.
- `'craig'` and `'satellite'` — the internal scale heuristic that drives
  `fitbounds` compares data bounds against world bounds in the projected
  plane, which is unreliable for these projections' non-linear (and,
  for satellite, perspective-dependent) mapping. Auto-fit would zoom
  incorrectly, so v4 skips it.

For these projections `layout.geo.fitbounds` is silently forced to `false`
in the resolved layout even if you leave it at the default. To fit
locations under one of these projections, supply the view attributes
yourself (`center`, `projection.rotation`, `projection.scale`, etc.).

### Reset button behavior

The modebar's Reset view button restores the auto-fit view when
`fitbounds` is active, not the schema default. Prior to v4 there was no
auto-fit default, so Reset simply restored the schema default; in v4 the

> initial view captured for reset is the fitted view.

Reviewer 2: I don't think this needs to be called out explicitly; this is how all of our axis autorange behavior is supposed to work.

```suggestion
```

gridlines at the same positions as its base axis, computing labels to
match — rather than picking its own auto-fit tick positions.

In v3 the two axes drew independent auto-tick grids, producing
staggered gridlines that were hard to read and left no visual
association between a gridline and a specific axis. Sync eliminates
that overlap.

```js
Plotly.newPlot(gd, [
    { y: [1, 3, 2], name: 'A' },
    { y: [100, 300, 200], name: 'B', yaxis: 'y2' }
], {
    yaxis2: { overlaying: 'y', side: 'right' }
    // v3: y2 picks its own ticks, producing a staggered grid
    // v4: y2 places ticks at y's tick positions, with y2-range labels
});
```

| Before (v3) | After (v4) |
|---|---|
| ![v3 staggered overlay ticks](/all_static/images/migrating-to-v4/migration_tickmode_v3.png) | ![v4 sync'd overlay ticks](/all_static/images/migrating-to-v4/migration_tickmode_v4.png) |

### Exception: categorical / multicategory base axes

When the overlaid (base) axis is `type: 'category'` or `'multicategory'`,
`tickmode` continues to default to `'auto'` on the overlaying axis.
Sync'ing tick positions to category slots is almost never the intent.

> ### Opting out

Reviewer 3: Should be a sub header

Set `tickmode: 'auto'` (or any other explicit value) on the overlaying
axis to restore v3 behavior:

```js
{ yaxis2: { overlaying: 'y', side: 'right', tickmode: 'auto' } }
```

> ### Incidental bugfix

Reviewer 4: ```suggestion
### Incidental bug fix
```
But making the heading more specific would be better
